crates/api/src/linker.rs

```diff
@@ -77,9 +81,40 @@ impl Linker {
             map: HashMap::new(),
             string2idx: HashMap::new(),
             strings: Vec::new(),
+            allow_shadow: false,
         }
     }
 
+    /// Configures whether this [`Linker`] will shadow previous duplicate
+    /// definitions of the same signature.
+    ///
+    /// By default a [`Linker`] will disallow duplicate definitions of the same
+    /// signature. This method, however, can be used to instead allow duplicates
+    /// and have the latest definition take precedence when linking modules.
```

Comment:
On most systems the symbols loaded first have priority I believe. (Otherwise `LD_PRELOAD` would be pretty much useless.)

Reply:
I agree that in the context of `LD_PRELOAD` that makes sense, but from from an API perspective, last-in-wins might be more intuitive for developers (a la virtual inheritance).  For example, I'd probably tell the linker to link in an engine-provided WASI implementation and then my host functions, which could potentially override the WASI implementation.

Reply:
`LD_PRELOAD`-style symbol preemption is fairly limited and awkward though. It's possible for libraries to make calls to their own symbols which can't be preempted, or to make assumptions that break if their symbols get preempted. So you have to depend on the library not doing those things.

I think it would be better to encourage wasm programs and tools to focus on interposition rather than preemption. Load the replacement *after* the thing it is replacing rather than before. One nice thing about this approach is that it means you can easy wrap functions and call them from your wrappers. For things like `malloc` replacements, it may require libraries to be structured differently, but it has the advantage of not depending on any implicit contracts with the libc to only access the malloc heap through the public interface.

Reply:
FWIW I think all behaviors are fine to implement. They're all simple enough and it's largely just a matter of defaults. If necessary it would be quite easy to implement a `LD_PRELOAD`-style first-one-always-wins, but I think it's best to wait for a use case to come up first.